warehouse/query-core/src/main/java/datawave/core/iterators/BoundedRangeExpansionIterator.java

```diff
@@ -0,0 +1,171 @@
+package datawave.core.iterators;
+
+import java.io.IOException;
+import java.util.Collection;
+import java.util.Map;
+import java.util.TreeSet;
+
+import org.apache.accumulo.core.data.ByteSequence;
+import org.apache.accumulo.core.data.Key;
+import org.apache.accumulo.core.data.Range;
+import org.apache.accumulo.core.data.Value;
+import org.apache.accumulo.core.iterators.IteratorEnvironment;
+import org.apache.accumulo.core.iterators.OptionDescriber;
+import org.apache.accumulo.core.iterators.SortedKeyValueIterator;
+import org.apache.accumulo.core.iterators.user.SeekingFilter;
+import org.apache.commons.lang3.StringUtils;
+import org.apache.hadoop.io.Text;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.base.Splitter;
+
+import datawave.query.Constants;
+import datawave.query.jexl.LiteralRange;
+
+/**
+ * A {@link SeekingFilter} that attempts to expand bounded ranges using the global index
+ * <p>
+ * The caller is responsible for fetching the appropriate column families. The range is constructed from a {@link LiteralRange}.
+ * <p>
+ * The only thing this iterator does is advance through datatypes if a filter is supplied, advance to the start date, and advance to the next row within the
+ * range.
+ */
+public class BoundedRangeExpansionIterator extends SeekingFilter implements OptionDescriber {
+
+    private static final Logger log = LoggerFactory.getLogger(BoundedRangeExpansionIterator.class);
+
+    public static final String START_DATE = "start.date";
+    public static final String END_DATE = "end.date";
+    public static final String DATATYPES_OPT = "dts";
+
+    private TreeSet<String> datatypes;
+    private String startDate;
+    private String endDate;
+
+    private Text prevRow;
+
+    @Override
+    public void init(SortedKeyValueIterator<Key,Value> source, Map<String,String> options, IteratorEnvironment env) throws IOException {
+        if (!validateOptions(options)) {
+            throw new IllegalArgumentException("BoundedRangeExpansionIterator not configured with correct options");
+        }
+
+        String opt = options.get(DATATYPES_OPT);
+        if (StringUtils.isBlank(opt)) {
+            datatypes = new TreeSet<>();
+        } else {
+            datatypes = new TreeSet<>(Splitter.on(',').splitToList(opt));
+        }
+
+        startDate = options.get(START_DATE);
+        endDate = options.get(END_DATE) + Constants.MAX_UNICODE_STRING;
+
+        super.init(source, options, env);
+    }
+
+    @Override
+    public IteratorOptions describeOptions() {
+        IteratorOptions opts = new IteratorOptions(getClass().getName(), "Expands bounded ranges using the global index", null, null);
+        opts.addNamedOption(START_DATE, "The start date");
+        opts.addNamedOption(END_DATE, "The end date");
+        opts.addNamedOption(DATATYPES_OPT, "The set of datatypes used to filter keys");
```

Comment:
Denote that the datatype option is optional

Reply:
done